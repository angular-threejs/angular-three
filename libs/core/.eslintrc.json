{
	"extends": ["../../.eslintrc.json"],
	"ignorePatterns": ["!**/*"],
	"overrides": [
		{
			"files": ["*.ts"],
			"extends": ["plugin:@nx/angular", "plugin:@angular-eslint/template/process-inline-templates"],
			"rules": {
				"@angular-eslint/directive-selector": "off",
				"@angular-eslint/component-selector": "off",
				"@angular-eslint/component-class-suffix": "off",
				"@angular-eslint/directive-class-suffix": "off",
				"@angular-eslint/no-host-metadata-property": "off",
				"@angular-eslint/no-input-rename": "off",
				"@typescript-eslint/no-explicit-any": "off",
				"@typescript-eslint/no-empty-function": "off",
        "@typescript-eslint/no-non-null-assertion": "off",
        "@typescript-eslint/ban-types": "off",
        "@typescript-eslint/no-empty-interface": "off",
        "@typescript-eslint/no-unused-vars": ["warn", { "argsIgnorePattern": "^_", "varsIgnorePattern": "^_" }]
			}
		},
    {
      "files": ["*.spec.ts"],
      "rules": {
        "@nx/enforce-module-boundaries": "off"
      }
    },
    {
      "files": ["web-gl-rendering-context.ts"],
      "rules": {
        "@typescript-eslint/ban-ts-comment": "off"
      }
    },
		{
			"files": ["*.html"],
			"extends": ["plugin:@nx/angular-template"],
			"rules": {}
		},
		{
			"files": ["*.json"],
			"parser": "jsonc-eslint-parser",
			"rules": {
				"@nx/dependency-checks": [
					"error",
					{
						"ignoredDependencies": [
							"@analogjs/vite-plugin-angular",
							"@angular/router",
							"@nx/devkit",
							"@nx/vite",
							"@phenomnomnominal/tsquery",
							"nx",
							"rxjs",
							"tslib",
							"vite"
						]
					}
				]
			}
		}
	]
}
