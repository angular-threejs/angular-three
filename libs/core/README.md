# `angular-three`

A custom Renderer for Angular 18+ that uses Three.js to render 3D scenes.

## Compatibility

Angular Three v2 is still in beta and aims to be compatible with Angular 17+.

## Installation

```bash
npm install angular-three@beta ngxtension three
# yarn add angular-three@beta ngxtension three
# pnpm add angular-three@beta ngxtension three
```

> Make sure to install `@types/three` as well

## Usage

```typescript
import { extend } from 'angular-three';
import { Mesh, BoxGeometry } from 'three';

extend({
	Mesh, // makes ngt-mesh available
	BoxGeometry, // makes ngt-box-geometry available
	/* ... */
	MyMesh: Mesh, // makes ngt-my-mesh available
});

// alternatively for demo purposes, you can use the following
// extend(THREE);
// This includes the entire THREE.js namespace

@Component({
	// This Component is rendered in the Custom Renderer
	standalone: true,
	template: `
		<ngt-mesh>
			<ngt-box-geometry />
		</ngt-mesh>
	`,
	schemas: [CUSTOM_ELEMENTS_SCHEMA], // required
	changeDetection: ChangeDetectionStrategy.OnPush,
})
export class SceneGraph {}

@Component({
	// This Component is rendered normally in Angular.
	selector: 'app-my-experience',
	standalone: true,
	template: `
		<ngt-canvas [sceneGraph]="SceneGraph" />
	`,
	imports: [NgtCanvas],
})
export class MyExperience {
	SceneGraph = SceneGraph;
}
```

> The Component that renders `NgtCanvas` (`MyExperience` in this case) controls the dimensions of the canvas so make sure to style it accordingly.

### Inputs

- `sceneGraph: Type<any>`: **required**. This is the Component that renders your 3D Scene graph. It must be a standalone Component.
- `gl?: NgtGLOptions`: This input allows you to configure the WebGL renderer used by Angular Three. You can provide a THREE.js renderer instance, properties for the default renderer, or a function that returns a renderer based on the canvas element.
- `size?: NgtSize`: Specifies the dimensions of the renderer. If omitted, the component will automatically measure the canvas dimensions.
- `shadows?: boolean | 'basic' | 'percentage' | 'soft' | 'variance' | Partial<WebGLShadowMap>`: Enables or disables shadows in the scene. You can provide a boolean value to toggle shadows on or off, or use specific strings to control the shadow type. Additionally, you can pass partial WebGLShadowMap options for fine-tuning.
- `legacy?: boolean`: Disables three r139 color management when set to true.
- `linear?: boolean`: Switches off automatic sRGB color space and gamma correction when set to true.
- `flat?: boolean`: Uses THREE.NoToneMapping instead of THREE.ACESFilmicToneMapping when set to true.
- `orthographic?: boolean`: Creates an orthographic camera instead of a perspective camera when set to true.
- `frameloop?: 'always' | 'demand' | 'never'`: Controls the rendering mode. 'always' renders continuously, 'demand' renders only on state changes, and 'never' gives you manual control over rendering.
- `performance?: Partial<Omit<NgtPerformance, 'regress'>>`: Allows you to configure performance options for adaptive performance.
- `dpr?: NgtDpr`: Sets the target pixel ratio. You can provide a single number or a range [min, max].
- `raycaster?: Partial<Raycaster>`: Configures the default raycaster used for interaction.
- `scene?: Scene | Partial<Scene>`: Provides a THREE.js scene instance or properties to create a default scene.
- `camera?: NgtCamera | Partial<NgtObject3DNode<Camera>>`: Provides a THREE.js camera instance or properties to create a default camera. You can also set the manual property to true to take control of the camera projection.
- `events?: (store: NgtSignalStore<NgtState>) => NgtEventManager<HTMLElement>`: Allows you to customize the event manager for handling pointer events.
- `eventSource?: HTMLElement | ElementRef<HTMLElement>`: Specifies the target element where events are subscribed. By default, it's the div wrapping the canvas.
- `eventPrefix?: 'offset' | 'client' | 'page' | 'layer' | 'screen'`: Sets the event prefix used for canvas pointer events.
- `lookAt?: Vector3 | Parameters<Vector3['set']>`: Defines the default coordinate for the camera to look at.

### Outputs

- `created`: Emitted when the canvas is created.
- `pointerMissed`: Emitted when a pointer event is not captured by any element (aka clicking on the canvas)

## Intellisense support

Since Angular Three is a custom Renderer, the elements are not recognized by the Angular Language Service.

### Jetbrains IDE

The consumers can add `web-types` property to the workspace's `package.json` and set the value to `node_modules/angular-three/web-types.json`.

```json
{
	"web-types": "node_modules/angular-three/web-types.json"
}
```

### VSCode

Similarly, there's `node_modules/angular-three/metadata.json` file that can be used to provide intellisense support for VSCode users.

The consumers can enable it via `html.customData` in their `settings.json` file.

```json
{
	"html.customData": ["node_modules/angular-three/metadata.json"]
}
```

## Input Bindings

Input bindings for `ngt-*` elements work the same way as they do in Angular.

> You can consult THREE.js documentation on what is available on the entities

```html
<ngt-mesh [position]="[x, y, z]" [rotation]="[x, y, z]">
	<ngt-mesh-basic-material color="hotpink" />
</ngt-mesh>
```

## Events

Angular Three Custom Renderer supports the following events on applicable objects (`ngt-mesh`, `ngt-group` etc...)

```
'click',
'contextmenu',
'dblclick',
'pointerup',
'pointerdown',
'pointerover',
'pointerout',
'pointerenter',
'pointerleave',
'pointermove',
'pointermissed',
'pointercancel',
'wheel',
```

In addition, there are 2 special events that the consumers can listen to;

- `attached`: when the element is attached to its parent
- `updated`: when the element properties are updated

## Constructor Arguments

In THREE.js, there are some entities that require the consumers to dispose and recreate them if their parameters change; like the Geometries.

To handle this, Angular Three exports a `NgtArgs` structural directive that always accepts an Array of values. The consumers can consult THREE.js documentations to know what values are applicable for what entities and their order.

```html
<!-- for example, new BoxGeometry(width, height, depth) -->
<ngt-box-geometry *args="[width, height, depth]" />
```

`NgtArgs`, as a structural directive, ensures to create a new instance of the entity when the value changes

## Parameters

Beside the normal properties that `ngt-*` elements can accept for Input bindings, the consumers can also pass a `parameters` object to a special property `[parameters]` on the elements. This parameters object will be used to apply the properties on the entity.

```html
<!-- instead of <ngt-mesh [position]="[x, y, z]" [scale]="scale" /> -->
<ngt-mesh [parameters]="{ position: [x, y, z], scale }" />
```

## Element Queries

The consumers can query for the THREE.js entities like they would do in normal HTML Angular Template.

```ts
@Component({
	template: `
		<ngt-mesh #mesh></ngt-mesh>
	`,
})
export class Box {
	mesh = viewChild.required<ElementRef<Mesh>>('mesh');
	//  notice that it is an ElementRef of THREE.Mesh instead of an HTMLElement
}
```

## Animation Loop

In order to participate in the animation loop, use `injectBeforeRender` inject function

```ts
@Component({
	/*...*/
})
export class Box {
	mesh = viewChild.required<ElementRef<Mesh>>('mesh');

	constructor() {
		injectBeforeRender(() => {
			// runs every frame
			const mesh = this.mesh().nativeElement;
			mesh.rotation.x += 0.01;
		});
	}
}
```

## Store

Angular Three keeps track of its state via an internal store. The consumers can access this store via the `injectStore` inject function

```ts
export class Box {
	store = injectStore();
	viewport = this.store.select('viewport'); // Signal<NgtViewport>
	camera = this.store.select('camera'); // Signal<NgtCamera> - the default camera
	/* many more properties */
}
```
