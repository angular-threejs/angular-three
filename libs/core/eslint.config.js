const { FlatCompat } = require('@eslint/eslintrc');
const js = require('@eslint/js');
const baseConfig = require('../../eslint.config.js');

const compat = new FlatCompat({
	baseDirectory: __dirname,
	recommendedConfig: js.configs.recommended,
});

module.exports = [
	...baseConfig,
	...compat
		.config({
			extends: ['plugin:@nx/angular', 'plugin:@angular-eslint/template/process-inline-templates'],
		})
		.map((config) => ({
			...config,
			files: ['**/*.ts'],
			files: ['**/*.ts'],
			rules: {
				'@angular-eslint/directive-selector': 'off',
				'@angular-eslint/component-selector': 'off',
				'@angular-eslint/component-class-suffix': 'off',
				'@angular-eslint/directive-class-suffix': 'off',
				'@angular-eslint/no-input-rename': 'off',
				'@typescript-eslint/no-explicit-any': 'off',
				'@typescript-eslint/no-empty-function': 'off',
				'@typescript-eslint/no-non-null-assertion': 'off',
				'@typescript-eslint/ban-types': 'off',
				'@typescript-eslint/no-empty-interface': 'off',
				'@typescript-eslint/no-unused-vars': [
					'warn',
					{
						argsIgnorePattern: '^_',
						varsIgnorePattern: '^_',
					},
				],
			},
		})),
	{
		files: ['**/*.spec.ts'],
		rules: {
			'@nx/enforce-module-boundaries': 'off',
		},
	},
	{
		files: ['web-gl-rendering-context.ts'],
		rules: {
			'@typescript-eslint/ban-ts-comment': 'off',
		},
	},
	...compat
		.config({
			extends: ['plugin:@nx/angular-template'],
		})
		.map((config) => ({
			...config,
			files: ['**/*.html'],
			rules: {
				...config.rules,
			},
		})),
	{
		files: ['**/*.json'],
		rules: {
			'@nx/dependency-checks': [
				'error',
				{
					ignoredDependencies: [
						'@analogjs/vite-plugin-angular',
						'@angular/router',
						'@nx/devkit',
						'@nx/vite',
						'@phenomnomnominal/tsquery',
						'nx',
						'rxjs',
						'tslib',
						'vite',
					],
				},
			],
		},
		languageOptions: {
			parser: require('jsonc-eslint-parser'),
		},
	},
	{
		files: ['**/*.ts'],
		rules: {
			'@angular-eslint/prefer-standalone': 'off',
		},
	},
];
