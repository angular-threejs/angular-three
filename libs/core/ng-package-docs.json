{
	"$schema": "../../node_modules/ng-packagr/ng-package.schema.json",
	"dest": "../../apps/docs/node_modules/angular-three",
	"lib": {
		"entryFile": "src/index.ts"
	},
	"allowedNonPeerDependencies": ["@nx/devkit", "nx", "@phenomnomnominal/tsquery"]
}
