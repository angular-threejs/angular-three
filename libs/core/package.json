{
	"name": "angular-three",
	"version": "0.0.0-replace",
	"publishConfig": {
		"access": "public"
	},
	"type": "module",
	"repository": {
		"type": "git",
		"url": "https://github.com/angular-threejs/angular-three/tree/main/libs/core"
	},
	"author": {
		"name": "Chau Tran",
		"email": "nartc7789@gmail.com",
		"url": "https://nartc.me"
	},
	"description": "Angular Renderer for THREE.js",
	"keywords": [
		"angular",
		"threejs",
		"renderer"
	],
	"license": "MIT",
	"peerDependencies": {
		"@angular/common": ">=18.0.0 <19.0.0",
		"@angular/core": ">=18.0.0 <19.0.0",
		"ngxtension": ">=3.0.0",
		"three": ">=0.148.0 <0.168.0",
    "node-three-gltf": ">=1.0.0 <2.0.0",
    "@nativescript/angular": ">=18.0.0 <19.0.0",
    "@nativescript/canvas": "2.0.0-webgpu.11",
    "@nativescript/canvas-three": "2.0.0-webgpu.11",
    "@nativescript/canvas-media": "2.0.0-webgpu.11"
  },
  "peerDependenciesMeta": {
    "node-three-gltf": {
      "optional": true
    },
    "@nativescript/angular": {
      "optional": true
    },
    "@nativescript/canvas": {
      "optional": true
    },
    "@nativescript/canvas-three": {
      "optional": true
    },
    "@nativescript/canvas-media": {
      "optional": true
    }
  },
	"dependencies": {
		"tslib": "^2.3.0",
		"@nx/devkit": "^19.0.0",
		"@phenomnomnominal/tsquery": "^6.0.0",
		"nx": "^19.0.0"
	},
	"generators": "./plugin/generators.json",
	"schematics": "./plugin/generators.json",
	"nx-migrations": {
		"migrations": "./plugin/migrations.json",
		"packageGroup": [
			"angular-three-soba",
			"angular-three-postprocessing",
			"angular-three-cannon",
			"angular-three-rapier"
		]
	},
	"ng-update": {
		"migrations": "./plugin/migrations.json",
		"packageGroup": [
			"angular-three-soba",
			"angular-three-postprocessing",
			"angular-three-cannon",
			"angular-three-rapier"
		]
	},
	"web-types": [
    "./web-types.json",
    "../../node_modules/angular-three/web-types.json"
  ]
}
