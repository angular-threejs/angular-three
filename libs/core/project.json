{
	"name": "core",
	"$schema": "../../node_modules/nx/schemas/project-schema.json",
	"sourceRoot": "libs/core/src",
	"prefix": "platform",
	"projectType": "library",
	"tags": [],
	"targets": {
		"build": {
			"executor": "@nx/angular:package",
			"outputs": ["{workspaceRoot}/dist/{projectRoot}"],
			"options": {
				"project": "libs/core/ng-package.json"
			},
			"configurations": {
				"production": {
					"tsConfig": "libs/core/tsconfig.lib.prod.json"
				},
				"development": {
					"tsConfig": "libs/core/tsconfig.lib.json"
				}
			},
			"defaultConfiguration": "production"
		},
		"build-ecosystem": {
			"executor": "nx:run-commands",
			"options": {
				"commands": ["pnpm exec nx build soba", "pnpm exec nx build cannon", "pnpm exec nx build postprocessing"],
				"parallel": true
			}
		},
		"package": {
			"executor": "nx:run-commands",
			"options": {
				"commands": [
					"pnpm exec nx build core",
					"pnpm exec nx build-ecosystem core",
					"pnpm exec nx generate-json core",
					"pnpm exec nx build plugin",
					"pnpm exec nx post-package core"
				],
				"parallel": false
			}
		},
		"generate-json": {
			"executor": "nx:run-commands",
			"options": {
				"commands": ["node ./tools/scripts/generate-json.mjs", "node ./tools/scripts/generate-soba-json.mjs"],
				"parallel": true
			}
		},
		"publish": {
			"command": "npm publish",
			"options": {
				"cwd": "dist/libs/core"
			}
		},
		"publish-beta": {
			"command": "npm publish --tag=beta",
			"options": {
				"cwd": "dist/libs/core"
			}
		},
		"post-package": {
			"executor": "nx:run-commands",
			"options": {
				"commands": ["node ./tools/scripts/release/post-package-core.mjs"],
				"parallel": false
			}
		},
		"test": {
			"executor": "@analogjs/vitest-angular:test"
		},
		"lint": {
			"executor": "@nx/eslint:lint",
			"outputs": ["{options.outputFile}"]
		}
	}
}
