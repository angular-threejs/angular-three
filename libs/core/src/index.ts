export * from './lib/canvas';
export * from './lib/directives/args';
export * from './lib/directives/selection';
export * from './lib/html';
export * from './lib/instance';
export * from './lib/loader';
export * from './lib/loop';
export * from './lib/pipes/hexify';
export * from './lib/portal';
export * from './lib/renderer';
export * from './lib/roots';
export * from './lib/routed-scene';
export * from './lib/store';
export * from './lib/utils/apply-props';
export * from './lib/utils/attach';
export * from './lib/utils/before-render';
export * from './lib/utils/is';
export * from './lib/utils/make';
export * from './lib/utils/non-nullish';
export * from './lib/utils/object-events';
export * from './lib/utils/parameters';
export * from './lib/utils/resolve-ref';
export * from './lib/utils/signal-store';
export * from './lib/utils/update';

export type * from './lib/three-types';
export type * from './lib/types';
