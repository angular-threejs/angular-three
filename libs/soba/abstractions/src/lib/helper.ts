import {
	ChangeDetectionStrategy,
	Component,
	computed,
	CUSTOM_ELEMENTS_SCHEMA,
	effect,
	ElementRef,
	Injector,
	input,
	signal,
	untracked,
	viewChild,
} from '@angular/core';
import { extend, getLocalState, injectBeforeRender, injectStore, resolveRef } from 'angular-three';
import { assertInjector } from 'ngxtension/assert-injector';
import { Object3D } from 'three';

type HelperArgs<T> = T extends [infer _, ...infer R] ? R : never;

export function injectHelper<
	TConstructor extends new (...args: any[]) => Object3D,
	THelperInstance extends InstanceType<TConstructor> & { update: () => void; dispose: () => void },
>(
	object: () => ElementRef<Object3D> | Object3D | undefined | null,
	helperConstructor: () => TConstructor,
	{
		injector,
		args = () => [] as unknown as HelperArgs<ConstructorParameters<TConstructor>>,
	}: { injector?: Injector; args?: () => HelperArgs<ConstructorParameters<TConstructor>> } = {},
) {
	return assertInjector(injectHelper, injector, () => {
		const store = injectStore();
		const scene = store.select('scene');

		const helper = signal<THelperInstance | null>(null);

		effect((onCleanup) => {
			let currentHelper: THelperInstance | undefined = undefined;
			const maybeObject3D = object();
			if (!maybeObject3D) return;

			const object3D = resolveRef(maybeObject3D);
			if (!object3D) return;

			currentHelper = new (helperConstructor() as any)(object3D, ...args());
			if (!currentHelper) return;

			untracked(() => {
				helper.set(currentHelper);
			});

			// Prevent the helpers from blocking rays
			currentHelper.traverse((child) => (child.raycast = () => null));
			scene().add(currentHelper);

			onCleanup(() => {
				helper.set(null);
				scene().remove(currentHelper);
				currentHelper.dispose?.();
			});
		});

		injectBeforeRender(() => {
			const currentHelper = helper();
			if (currentHelper) currentHelper.update();
		});

		return helper.asReadonly();
	});
}

@Component({
	standalone: true,
	selector: 'ngts-helper',
	template: `
		<ngt-object3D #helper />
	`,
	schemas: [CUSTOM_ELEMENTS_SCHEMA],
	changeDetection: ChangeDetectionStrategy.OnPush,
})
export class NgtsHelper<TConstructor extends new (...args: any[]) => Object3D> {
	type = input.required<TConstructor>();
	options = input<HelperArgs<ConstructorParameters<TConstructor>>>(
		[] as unknown as HelperArgs<ConstructorParameters<TConstructor>>,
	);

	helperRef = viewChild.required<ElementRef<Object3D>>('helper');

	private parent = computed(() => {
		const helper = this.helperRef().nativeElement;
		if (!helper) return;
		const localState = getLocalState(helper);
		if (!localState) return;
		return localState.parent();
	});

	helper = injectHelper(this.parent, this.type, { args: this.options });

	constructor() {
		extend({ Object3D });
	}
}
