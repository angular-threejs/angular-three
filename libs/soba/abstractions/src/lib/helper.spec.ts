import { ChangeDetectionStrategy, Component, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { NgtTestBed } from 'angular-three/testing';
import { BoxHelper } from 'three';
import { NgtsHelper } from './helper';

describe(NgtsHelper.name, () => {
	@Component({
		standalone: true,
		template: `
			<ngt-mesh>
				<ngt-box-geometry />
				<ngts-helper [type]="BoxHelper" [options]="$any(['royalblue'])" />
			</ngt-mesh>
		`,
		schemas: [CUSTOM_ELEMENTS_SCHEMA],
		changeDetection: ChangeDetectionStrategy.OnPush,
		imports: [NgtsHelper],
	})
	class SceneGraph {
		protected readonly BoxHelper = BoxHelper;
	}

	it('should render properly', async () => {
		const { scene, fixture, toGraph } = NgtTestBed.create(SceneGraph);
		fixture.detectChanges();

		expect(scene.children.length).toEqual(2);
		expect(scene.children[1].type).toEqual('BoxHelper');
		expect(toGraph()).toMatchSnapshot();
	});
});
