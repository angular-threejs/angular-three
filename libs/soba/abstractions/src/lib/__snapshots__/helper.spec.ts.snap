// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NgtsHelper > should render properly 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "name": "",
        "type": "Object3D",
      },
    ],
    "name": "",
    "type": "Mesh",
  },
  {
    "children": [],
    "name": "",
    "type": "BoxHelper",
  },
]
`;
